### PR TITLE
Use `get` instead of `join` to avoid getting stuck

### PR DESCRIPTION
**Pull request by 315157973**
### Motivation
 join() will not time out, when network jitter occurs, it is easy to cause the thread to get stuck forever

### Modifications
Use get(timeout) instead of join() to avoid getting stuck

### CODE REVIEW

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/OwnershipCache.java [hunk 1]
```diff
@@ -258,7 +261,12 @@ public OwnedBundle getOwnedBundle(NamespaceBundle bundle) {
         CompletableFuture<OwnedBundle> future = ownedBundlesCache.getIfPresent(bundle);
 
         if (future != null && future.isDone() && !future.isCompletedExceptionally()) {
-            return future.join();
+            try {
+                return future.get(pulsar.getConfiguration().getZooKeeperOperationTimeoutSeconds(), TimeUnit.SECONDS);
+            } catch (Exception e) {
+                log.error("Get owned bundle failed ", e);
+                return null;
```
Comment (eolivelli) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/OwnershipCache.java:268]: This is a behaviour change.
We should fail and not return null

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerBuilderImpl.java [hunk 2]
```diff
@@ -122,11 +122,17 @@ public ConsumerBuilderImpl(PulsarClientImpl client, Schema<T> schema) {
...
+            try {
+                if (client.getPartitionedTopicMetadata(oldRetryLetterTopic)
+                        .get(client.conf.getOperationTimeoutMs(), TimeUnit.MILLISECONDS).partitions > 0) {
+                    retryLetterTopic = oldRetryLetterTopic;
+                }
+                if (client.getPartitionedTopicMetadata(oldDeadLetterTopic)
+                        .get(client.conf.getOperationTimeoutMs(), TimeUnit.MILLISECONDS).partitions > 0) {
+                    deadLetterTopic = oldDeadLetterTopic;
+                }
+            } catch (Exception e) {
```
Comment (eolivelli) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerBuilderImpl.java:134]: Can we catch the specific exception? This is a new 'catch' block

**Review by eolivelli — CHANGES_REQUESTED** [review]
Overall the patch LGTM

But I left  few comments

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/OwnershipCache.java (hunk 1 above)
Comment (315157973) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/OwnershipCache.java]: Nice view, done

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerBuilderImpl.java (hunk 2 above)
Comment (315157973) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerBuilderImpl.java]: done

**Review by eolivelli — APPROVED** [review]
Lgtm

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerBuilderImpl.java [hunk 3]
```diff
@@ -122,11 +124,17 @@ public ConsumerBuilderImpl(PulsarClientImpl client, Schema<T> schema) {
...
+                if (client.getPartitionedTopicMetadata(oldRetryLetterTopic)
+                        .get(client.conf.getOperationTimeoutMs(), TimeUnit.MILLISECONDS).partitions > 0) {
+                    retryLetterTopic = oldRetryLetterTopic;
+                }
+                if (client.getPartitionedTopicMetadata(oldDeadLetterTopic)
+                        .get(client.conf.getOperationTimeoutMs(), TimeUnit.MILLISECONDS).partitions > 0) {
+                    deadLetterTopic = oldDeadLetterTopic;
+                }
+            } catch (InterruptedException | TimeoutException | ExecutionException e) {
+                return FutureUtil.failedFuture(e);
```
Comment (codelipenghui) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerBuilderImpl.java:137]: Should return FutureUtil.failedFuture(e.getCause()) if e is ExecutionException?

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/v1/NonPersistentTopics.java [hunk 4]
```diff
@@ -297,7 +298,14 @@ public void getList(@Suspended final AsyncResponse asyncResponse, @PathParam("pr
...
 
     private Topic getTopicReference(TopicName topicName) {
-        return pulsar().getBrokerService().getTopicIfExists(topicName.toString()).join()
-                .orElseThrow(() -> new RestException(Status.NOT_FOUND, "Topic not found"));
+        try {
+            return pulsar().getBrokerService().getTopicIfExists(topicName.toString())
+                    .get(config().getZooKeeperOperationTimeoutSeconds(), TimeUnit.SECONDS)
+                    .orElseThrow(() -> new RestException(Status.NOT_FOUND, "Topic not found"));
+        } catch (ExecutionException e) {
+            throw new RuntimeException(e.getCause());
```
Comment (codelipenghui) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/v1/NonPersistentTopics.java:306]: Should throw RestException here?

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerBuilderImpl.java (hunk 3 above)
Comment (315157973) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerBuilderImpl.java:137]: Done

**Review by gaoran10 — APPROVED** [review]
LGTM

**Review by codelipenghui — APPROVED** [review]

**Comment by hangc0276** [comment]
This PR depend on #11265 , which is a feature. we move this pr to 2.8.2 first.

**Pull request merged**